Make FieldSplit preconditioner compatible with mortar contact

dewenyushu opened the pull request:
With this fix, the FieldSplit converges for frictional mortar contact problems. This addresses part of the issue #14791.

<!--
INCLUDE THE FOLLOWING IN THE PR DESCRIPTION
- Explain relevant design information for your change.
- Follow the [Coding Standards](http://mooseframework.org/wiki/CodeStandards/).
- Submit or improve [Test Cases](http://mooseframework.org/wiki/MooseTraining/testing/).
- Reference a specific issue, place "refs #<issue>" or "closes #<issue>" (e.g., #closes #1234).
-->

dewenyushu committed:
Make FieldSplit preconditioner compatible with mortar contact, issue #14791

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/470646/) on a5efe9c wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/14792/site)

This comment will be updated on new commits.

dewenyushu commented:
@bwspenc please review this PR which fixes the FieldSplit preconditioner issue with mortar. Thanks!

fdkong commented on framework/src/utils/PetscDMMoose.C:
```diff
@@ -743,6 +743,19 @@ DMMooseGetEmbedding_Private(DM dm, IS * embedding)
           }
+          // loop over lower dimensional slave elements
+          for (const auto & bit : *(dmm->_block_names))
+            if (bit.second.find("_slave_subdomain") != std::string::npos)
```
The subdomain name should not be hard-coded.  @lindsayad  Did you use this name by default when you creating a subdomain for mortar?

lindsayad replied:
This `DMMoose` stuff is hard-coded to work with a `Contact` block. And yes in `Contact` we automatically set up the lower-dimensional subdomains with these guaranteed subdomains. I actually voiced this exact same concern to @dewenyushu, but she correctly pointed out this fact to me. If we generalize `DMMoose` to work with arbitrary constraints then we won't be able to do what she's doing here, and we'll have to add some kind of map to `MooseMesh` for retrieving lower-d subdomains associated with sidesets. But for now I think this is ok.

fdkong replied:
Yes, It may be the time to refactor DMMoose to take arbitrary blocks. The basic idea is simple here: we just "abuse" FS to mimic block elimination.  

It would be helpful if you @lindsayad could add API in MooseMesh so that we could retrieve this info easily. Hard-coded thing is always bad.

We could leave this PR as is, but we should make these changes in the future

lindsayad replied:
Yes, I think that that is beyond @dewenyushu's scope here. I can add that `MooseMesh` support after this goes in, and use her tests as a bellwether for that new support

fdkong replied:
Great

fdkong commented on framework/src/utils/PetscDMMoose.C:
```diff
@@ -743,6 +743,19 @@ DMMooseGetEmbedding_Private(DM dm, IS * embedding)
+            if (bit.second.find("_slave_subdomain") != std::string::npos)
+              for (const auto & elem : as_range(
+                       dmm->_nl->system().get_mesh().active_subdomain_elements_begin(bit.first),
+                       dmm->_nl->system().get_mesh().active_subdomain_elements_end(bit.first)))
```
Please use `active_local_subdomain_elements_begin`

fdkong commented on modules/contact/test/tests/fieldsplit/frictional_mortar_FS.i:
```diff
@@ -0,0 +1,183 @@
+      include_all_contact_nodes = 1
+
+      petsc_options_iname = '-ksp_type -pc_type -pc_hypre_type '
+      petsc_options_value = '  preonly hypre  boomeramg'
```
HYPRE is more efficient when you use it as a preconditioner.

lindsayad replied:
Huh? Isn't that what she's doing?

fdkong replied:
I meant: inside of FS, you still can use KSP combined with PC that make the entire solver more robust. In some cases, we may even need to do iteration between blocks.

dewenyushu replied:
> I meant: inside of FS, you still can use KSP combined with PC that make the entire solver more robust.

Interesting.. I am going to try using KSP inside FS.

>  In some cases, we may even need to do iteration between blocks.

 Isn't FS inherently iterating between blocks?

fdkong replied:
One iteration only right now :)

fdkong replied:
> Interesting.. I am going to try using KSP inside FS

Remember to set the number of iterations as 2 or 3 ... when try this. It is problem-dependent. But do not worry about this at this point

dewenyushu replied:
@fdkong Now I see what you mean here... Yea I agree including more iterations would definitely improve the robustness of the preconditioner. Thanks for pointing this out!

fdkong commented on modules/contact/test/tests/fieldsplit/tests:
```diff
@@ -1,11 +1,19 @@
+    input = 'frictional_mortar_FS.i'
+    exodiff = 'frictional_mortar_FS_out.e'
+    abs_zero = 1e-8
+    max_parallel = 1
```
Is there a particular reason that the code does not run in parallel?

dewenyushu replied:
It did not pass some of the parallel moosebuild tests so I made this to run only in serial. I found the test error was due to some tolerance settings, and I will update the input and tests file correspondingly.

lindsayad replied:
Yea I would definitely hope that this field split capability allows us to run in parallel otherwise I see little advantage over using LU (which definitely works in parallel)

lindsayad replied:
> It did not pass some of the parallel moosebuild tests so I made this to run only in serial. I found the test error was due to some tolerance settings, and I will update the input and tests file correspondingly.

👍

dewenyushu replied:
> Yea I would definitely hope that this field split capability allows us to run in parallel otherwise I see little advantage over using LU (which definitely works in parallel)

Yea I agree! Just FYI, I ran some cases to compare the performances of different preconditioners for mortar contact problems (in serial). It looks like FS performs better than LU for relatively small problem sizes, but cannot compete super LU (not very surprising...). I would suppose FS performs and scales better for large problems in parallel, and that would be my next task.

fdkong replied:
Better in terms of compute time or iteration?

dewenyushu replied:
> Better in terms of compute time or iteration?

FS is better in terms of total Flop.

lindsayad replied:
Yea, so by what metric is LU better?

dewenyushu replied:
> Yea, so by what metric is LU better?

I choose Flop as the only metric for now (should be consistent with compute time). Also, I only have serial comparisons. Under this setting, FS performs better than LU (but not superLU) for problem sizes around 1e3 elements. However, I have yet to examine the parallel performance. The hope is that FS -- as an iterative preconditioner -- should perform much better than LU (hopefully superLU, too) when run in parallel. I will keep you posted on this :)

lindsayad replied:
> Under this setting, FS performs better than LU (but not superLU) for problem sizes around 1e3 elements

So when running in serial, for 1000 elements, performance is like this: SuperLU is better than FSP which is better than PETSc default LU?

lindsayad replied:
Could you try running with like 100,000 elements (in serial)? @fdkong and I hope that at that point FSP should be more performant than SuperLU. If you want, you can come visit the MOOSE cube. We are both here this week

dewenyushu replied:
> So when running in serial, for 1000 elements, performance is like this: SuperLU is better than FSP which is better than PETSc default LU?

Yes, that is correct for mortar contact.

dewenyushu replied:
> Could you try running with like 100,000 elements (in serial)? @fdkong and I hope that at that point FSP should be more performant than SuperLU. If you want, you can come visit the MOOSE cube. We are both here this week.
> 
> I don't know how many time steps you're running...100,000 elements in serial may take too long...

I would expect 1e5 elements being too large for a serial run. This is also why I would like to try comparing parallel performances. I am currently occupied with something else, but yea, I will visit you guys later this week in order to move this forward!

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/471270/) on 265372e wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/14792/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/14792/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format acf3591472bba3582d6581d12eb631880eabe27f`

fdkong commented:
I am looking at the` PETSc submodule` failing

fdkong committed:
FSP: Separate petscSetupDM() from petscSetOptions

We need to call petscSetupDM every FEProblemBase::solve()
since libmesh destories SNES after solve()

fdkong committed:
FSP: get rid of hard-coded subdomain name

Users should just use the input  parameter "blokcs" to specify what they want to include

fdkong commented:
I am able to get rid of the hard-coded subdomain name, and #14804 is not a priority anymore. @lindsayad

lindsayad commented on modules/contact/test/tests/fieldsplit/tests:
```diff
@@ -1,11 +1,18 @@
-  issues = '#7485'
-  design = 'FieldSplitPreconditioner.md'
+  issues = '#7485 #14791'
+  design = 'FieldSplitPreconditioner.md NormalMortarLMMechanicalContact.md TangentialMortarLMMechanicalContact.md'
```
This should be NormalNodalLMMechanicalContact

lindsayad reviewed (changes requested):
@dewenyushu do you want to explore the more robust linear solve like we talked about today in this PR, or are you satisfied with the test cases here?

lindsayad commented on modules/contact/test/tests/fieldsplit/tests:
```diff
@@ -1,11 +1,18 @@
+    input = 'frictional_mortar_FS.i'
+    exodiff = 'frictional_mortar_FS_out.e'
+    abs_zero = 1e-8
+    requirement = 'The mortar contact problem shall converge reasonably fast with split preconditioning.'
```
“The system shall allow split preconditioning of frictional mortar contact.”

I don’t think we’re prepared to comment on its efficiency yet.

dewenyushu replied:
> “The system shall allow split preconditioning of frictional mortar contact.”
> 
> I don’t think we’re prepared to comment on its efficiency yet.

Yea I agree. This example shows FS works with mortar contact. The efficiency is yet to be investigated and improved.

dewenyushu commented:
> @dewenyushu do you want to explore the more robust linear solve like we talked about today in this PR, or are you satisfied with the test cases here?

The linear solve is to be improved. But I suppose that does not necessarily have to be in this PR, since:
 i) this PR mainly shows FS now converges for mortar (which it used to diverge without this fix ) 
 ii) the linear solve issue exhibits for all other preconditioners/solvers and is probably more reasonable to be addressed in another PR

dewenyushu commented on framework/src/utils/PetscDMMoose.C:
```diff
@@ -743,6 +743,19 @@ DMMooseGetEmbedding_Private(DM dm, IS * embedding)
                 indices.insert(edof);
           }
+          // loop over lower dimensional slave elements
+          for (const auto & bit : *(dmm->_block_ids))
```
I am confused.. does this loop through all the elements, including the ones not on the slave?

dewenyushu committed:
Fix tests file for FS

dewenyushu commented:
> I am able to get rid of the hard-coded subdomain name

Great job! Glad to know that we can get rid of the default lower-dimensional block names and the PETSc issue has been fixed
